src/DynamoCoreWpf/ViewModels/PackageManager/PackageManagerSearchElementViewModel.cs

```diff
@@ -41,6 +50,46 @@ public PackageManagerSearchElementViewModel(PackageManagerSearchElement element,
                 new DelegateCommand(() => GoToUrl(FormatUrl(Model.SiteUrl)), () => !String.IsNullOrEmpty(Model.SiteUrl));
             this.VisitRepositoryCommand =
                 new DelegateCommand(() => GoToUrl(FormatUrl(Model.RepositoryUrl)), () => !String.IsNullOrEmpty(Model.RepositoryUrl));
+
+        }
+
+        /// <summary>
+        /// Alternative constructor which takes a Function object in order to
+        /// assist communication between the PackageManagerSearchViewModel
+        /// and the PackageManagerSearchElementViewModel.
+        /// </summary>
+        /// <param name="element">A PackageManagerSearchElement</param>
+        /// <param name="canLogin">A Boolean used for access control to certain internal packages.</param>
+        /// <param name="isPackageInstalledFunction">A function that we pass to this constructor which takes a string and returns a Boolean.</param>
+        public PackageManagerSearchElementViewModel(PackageManagerSearchElement element, bool canLogin, Func<string,bool> isPackageInstalledFunction) : this(element, canLogin)
+        {
+            this.isPackageInstalled = isPackageInstalledFunction;
+            this.IsInstalled = isPackageInstalled(this.Model.Name);
+        }
+
+        private bool isInstalled;
```

Comment:
I think we should use the existing `PackageLoadState` APIs to tell if a user has already "installed" a package or not.

When a Package is successfully downloaded and Loaded into Dynamo it gets assigned a "LoadState.State == ackageLoadState.StateTypes.Loaded"
If a package is downloaded but Loading into Dynamo fails...it gets assigned "LoadState.State == ackageLoadState.StateTypes.Error"
If a package is downloaded but intentionally not Loaded into dynamo it gets assigned "LoadState.State == ackageLoadState.StateTypes.Unloaded"
Another thing to keep in mind is that a Package can be Built-in and in one of the states above, but I guess a user should be able to install a non-built-in package with the same name

@OliverEGreen @mjkkirschner @QilongTang 
When should the user be able to install/reinstall a package xxx with version yyy?
1. The package xxx@yyy does not exist (i.e not in the LocalPackages) - Yes
2. The package xxx@yyy is in an error state - Yes
3. The package xxx@yyy is loaded but built-in - Yes
4. The package xxx@yyy is unloaded - No
5. The pacakge xxx@yyy has a state == None (i.e downloaded but not processed by Dynamo yet) - Not sure